### ISSUE TITLE
matplotlib.pyplot import gives error on MacOSX

### ISSUE THREAD

**Issue by lucilla7** [matplotlib/matplotlib#11822]
### Bug report

**Bug summary**

Import matplotlib.pyplot fails on MacOSX

**Code for reproduction**

```python
import matplotlib.pyplot as plt
```

**Actual outcome**

```
Traceback (most recent call last):
  File "/Users/Apollo/PycharmProjects/CT_registration/main.py", line 114, in <module>
    moving_npa=fixed(sitk.GetArrayViewFromImage(movingImage)))
  File "/Users/Apollo/venv/lib/python2.7/site-packages/ipywidgets/widgets/interaction.py", line 531, in __call__
    show_inline_matplotlib_plots()
  File "/Users/Apollo/venv/lib/python2.7/site-packages/ipywidgets/widgets/interaction.py", line 52, in show_inline_matplotlib_plots
    from ipykernel.pylab.backend_inline import flush_figures
  File "/Users/Apollo/venv/lib/python2.7/site-packages/ipykernel/pylab/backend_inline.py", line 9, in <module>
    from matplotlib.backends.backend_agg import new_figure_manager, FigureCanvasAgg # analysis: ignore
  File "/Users/Apollo/venv/lib/python2.7/site-packages/matplotlib/backends/backend_agg.py", line 32, in <module>
    from matplotlib.backend_bases import (
  File "/Users/Apollo/venv/lib/python2.7/site-packages/matplotlib/backend_bases.py", line 53, in <module>
    from matplotlib import (
  File "/Users/Apollo/venv/lib/python2.7/site-packages/matplotlib/textpath.py", line 17, in <module>
    import matplotlib.font_manager as font_manager
  File "/Users/Apollo/venv/lib/python2.7/site-packages/matplotlib/font_manager.py", line 1469, in <module>
    _rebuild()
  File "/Users/Apollo/venv/lib/python2.7/site-packages/matplotlib/font_manager.py", line 1450, in _rebuild
    fontManager = FontManager()
  File "/Users/Apollo/venv/lib/python2.7/site-packages/matplotlib/font_manager.py", line 1086, in __init__
    self.ttflist = createFontList(self.ttffiles)
  File "/Users/Apollo/venv/lib/python2.7/site-packages/matplotlib/font_manager.py", line 587, in createFontList
    font = ft2font.FT2Font(fpath)
TypeError: First argument must be a path or file object reading bytes

Process finished with exit code 1
```

**Expected outcome**

**Matplotlib version**

  * Operating system: Mac OS X Lion 10.7.5
  * Matplotlib version: 2.2.2
  * Matplotlib backend: MacOSX
  * Python version: 2.7.1
  * Jupyter version: jupyter-client==5.2.3 jupyter-core==4.4.0
  * Other libraries: SimpleITK, ipywidgets

I installed matplotlib from source. I am on a virtual environment.

**Comment by ImportanceOfBeingErnest** [comment]
The error does not fit to the code line reported. I would suggest to show the minimal code that reproduces the problem and the traceback that this minimal code shows.

**Comment by lucilla7** [comment]
> Traceback (most recent call last):
  File "/Users/Apollo/PycharmProjects/CT_registration/error.py", line 1, in <module>
    import matplotlib.pyplot as plt
  File "/Users/Apollo/venv/lib/python2.7/site-packages/matplotlib/pyplot.py", line 31, in <module>
    import matplotlib.colorbar
  File "/Users/Apollo/venv/lib/python2.7/site-packages/matplotlib/colorbar.py", line 36, in <module>
    import matplotlib.contour as contour
  File "/Users/Apollo/venv/lib/python2.7/site-packages/matplotlib/contour.py", line 20, in <module>
    import matplotlib.font_manager as font_manager
  File "/Users/Apollo/venv/lib/python2.7/site-packages/matplotlib/font_manager.py", line 1469, in <module>
    _rebuild()
  File "/Users/Apollo/venv/lib/python2.7/site-packages/matplotlib/font_manager.py", line 1450, in _rebuild
    fontManager = FontManager()
  File "/Users/Apollo/venv/lib/python2.7/site-packages/matplotlib/font_manager.py", line 1086, in __init__
    self.ttflist = createFontList(self.ttffiles)
  File "/Users/Apollo/venv/lib/python2.7/site-packages/matplotlib/font_manager.py", line 587, in createFontList
    font = ft2font.FT2Font(fpath)
TypeError: First argument must be a path or file object reading bytes


This is the traceback generated by

import matplotlib.pyplot as plt

**Comment by WeatherGod** [comment]
You are running in a pycharm environment, aren't you? I wonder if the
matplotlib it is loading isn't the one you installed?

**Comment by lucilla7** [comment]
Yes, I'm using PyCharm.

Running this couple of lines:
`import matplotlib`
`print matplotlib.__version__`

I get this output:

> 2.2.2

The matplotlib loaded is the same I have installed

**Comment by jklymak** [comment]
Can you check if this works outside PyCharm?  I expect it does, or we'd have lots more bug reports.  If so, I think the issue is either with how you've setup pycharm, or an issue with how pycharm resolves paths.

**Comment by QuLogic** [comment]
> The error does not fit to the code line reported. I would suggest to show the minimal code that reproduces the problem and the traceback that this minimal code shows.

What do you mean? [This line](https://github.com/matplotlib/matplotlib/blob/v2.2.2/lib/matplotlib/font_manager.py#L587) is consistent with what should be in 2.2.2.

**Comment by ImportanceOfBeingErnest** [comment]
@QuLogic The original post has a traceback which starts in line 114 of some code, that cannot be the one liner `import matplotlib.pyplot as plt`. To make sure the issue is actually with matplotlib and to find its cause one needs to see the traceback that is produced by the reported code. The OP reacted in an exemplary way to this comment and provided the [correct code and traceback](https://github.com/matplotlib/matplotlib/issues/11822#issuecomment-411072294) 7 minutes later.

**Comment by tacaswell** [comment]
I think this is a duplicate of https://github.com/matplotlib/matplotlib/issues/11635 (which is fixed on 2.2.x and master).

This makes some sense as macOS is BSD derived, but what confuses me is that we have not seen many more reports of this and our many OSX-based developers have not seen it.

**Comment by jklymak** [comment]
I havent used 2.7 since before 2.2

**Comment by lucilla7** [comment]
`source venv/bin/activate
(venv)Mac-Pro-di-Apollo:~ Apollo$ python
Python 2.7.1 (r271:86832, Jul 31 2011, 19:30:53) 
[GCC 4.2.1 (Based on Apple Inc. build 5658) (LLVM build 2335.15.00)] on darwin
Type "help", "copyright", "credits" or "license" for more information.

>>> import matplotlib.pyplot as plt
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/Apollo/venv/lib/python2.7/site-packages/matplotlib/pyplot.py", line 31, in <module>
    import matplotlib.colorbar
  File "/Users/Apollo/venv/lib/python2.7/site-packages/matplotlib/colorbar.py", line 36, in <module>
    import matplotlib.contour as contour
  File "/Users/Apollo/venv/lib/python2.7/site-packages/matplotlib/contour.py", line 20, in <module>
    import matplotlib.font_manager as font_manager
  File "/Users/Apollo/venv/lib/python2.7/site-packages/matplotlib/font_manager.py", line 1469, in <module>
    _rebuild()
  File "/Users/Apollo/venv/lib/python2.7/site-packages/matplotlib/font_manager.py", line 1450, in _rebuild
    fontManager = FontManager()
  File "/Users/Apollo/venv/lib/python2.7/site-packages/matplotlib/font_manager.py", line 1086, in __init__
    self.ttflist = createFontList(self.ttffiles)
  File "/Users/Apollo/venv/lib/python2.7/site-packages/matplotlib/font_manager.py", line 587, in createFontList
    font = ft2font.FT2Font(fpath)
TypeError: First argument must be a path or file object reading bytes`

This is the traceback from terminal
I add the also the results from pip freeze:

`absl-py==0.1.13
amico==1.0
appnope==0.1.0
attrs==17.4.0
backports-abc==0.5
backports.functools-lru-cache==1.5
backports.shutil-get-terminal-size==1.0.0
backports.ssl-match-hostname==3.5.0.1
backports.weakref==1.0.post1
bleach==1.5.0
certifi==2017.7.27
click==6.7
cloudpickle==0.5.3
configparser==3.5.0
cycler==0.10.0
Cython==0.28.1
dask==0.18.2
decorator==4.2.1
dipy==0.13.0
entrypoints==0.2.3
enum34==1.1.6
funcsigs==1.0.2
future==0.16.0
futures==3.2.0
h5py==2.7.1
html5lib==0.9999999
ipykernel==4.8.2
ipython==5.8.0
ipython-genutils==0.2.0
ipywidgets==7.4.0
isodate==0.6.0
Jinja2==2.10
jsonschema==2.4.0
jupyter-client==5.2.3
jupyter-core==4.4.0
kiwisolver==1.0.1
lxml==4.2.1
mahotas==1.4.4
Markdown==2.6.11
MarkupSafe==1.0
matplotlib==2.2.2
mistune==0.8.3
mock==2.0.0
more-itertools==4.1.0
mpmath==1.0.0
nbconvert==5.3.1
nbformat==4.4.0
networkx==2.1
nibabel==2.2.1
nilearn==0.4.1
nipy==0.4.2
nipype==1.0.2
notebook==5.0.0
numpy==1.14.2
packaging==17.1
pandocfilters==1.4.2
pathlib2==2.3.2
pbr==4.0.0
pexpect==4.6.0
pickleshare==0.7.4
Pillow==5.2.0
pluggy==0.6.0
prometheus-client==0.3.1
prompt-toolkit==1.0.15
protobuf==3.5.2.post1
prov==1.5.0
ptyprocess==0.6.0
py==1.5.3
pydot==1.2.4
pydotplus==2.0.2
Pygments==2.2.0
pyparsing==2.2.0
pytest==3.5.0
python-dateutil==2.7.2
pytz==2018.3
PyWavelets==0.5.2
pyzmq==17.1.0
rdflib==4.2.2
scandir==1.8
scikit-image==0.14.0
scikit-learn==0.19.1
scipy==1.0.1
simplegeneric==0.8.1
SimpleITK==1.1.0
simplejson==3.13.2
singledispatch==3.4.0.3
six==1.11.0
sklearn==0.0
spams==2.6.1
subprocess32==3.2.7
sympy==1.1.1
tensorflow==1.5.0
tensorflow-tensorboard==1.5.1
terminado==0.8.1
testpath==0.3.1
toolz==0.9.0
tornado==4.3
traitlets==4.3.2
traits==4.6.0
wcwidth==0.1.7
Werkzeug==0.14.1
widgetsnbextension==3.4.0`

**Comment by tacaswell** [comment]
@lucilla7 does any version of 2.2.x work for you on osx?

This looks like it is a bug in some low-level c code so the question is why has this been working on travis at all (ex https://travis-ci.org/matplotlib/matplotlib/jobs/395614611 should be failing).

Can one of the developers with a mac try to reproduce this?  We current have 2 fixes (which are redundant with each other) one merged (#11636) and one proposed (#11827) so I am not worried about this being fixed for 2.2.3, but it would be good to document which versions of 2.2.x do not work on py27 + OSX.

**Comment by tacaswell** [comment]
You _may_ have to blow away your font cache to trigger this?

**Comment by dopplershift** [comment]
I cannot reproduce this problem on OSX with matplotlib 2.2.2 from conda-forge, even if I blow away my font cache.

**Comment by danielballan** [comment]
Also cannot reproduce on OSX:

```
$ python --version
Python 2.7.15 :: Anaconda, Inc.
```

```
$ env | grep MPLBACKEND
MPLBACKEND=tkagg
```

```
$ python -c "import matplotlib.pyplot; import matplotlib; print(matplotlib.__version__)"
2.2.0
```

```
$ python -c "import matplotlib.pyplot; import matplotlib; print(matplotlib.__version__)"
2.2.2
```

**Comment by lucilla7** [comment]
I tried with another virtual environment based on python 2.7.9, and it is the same: 

`(venv2714)Mac-Pro-di-Apollo:~ Apollo$ python`
`Python 2.7.9 (v2.7.9:648dcafa7e5f, Dec 10 2014, 10:10:46) `
`[GCC 4.2.1 (Apple Inc. build 5666) (dot 3)] on darwin`

>>> import matplotlib.pyplot as plt
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/Apollo/venv2714/lib/python2.7/site-packages/matplotlib/pyplot.py", line 31, in <module>
    import matplotlib.colorbar
  File "/Users/Apollo/venv2714/lib/python2.7/site-packages/matplotlib/colorbar.py", line 36, in <module>
    import matplotlib.contour as contour
  File "/Users/Apollo/venv2714/lib/python2.7/site-packages/matplotlib/contour.py", line 20, in <module>
    import matplotlib.font_manager as font_manager
  File "/Users/Apollo/venv2714/lib/python2.7/site-packages/matplotlib/font_manager.py", line 1469, in <module>
    _rebuild()
  File "/Users/Apollo/venv2714/lib/python2.7/site-packages/matplotlib/font_manager.py", line 1450, in _rebuild
    fontManager = FontManager()
  File "/Users/Apollo/venv2714/lib/python2.7/site-packages/matplotlib/font_manager.py", line 1086, in __init__
    self.ttflist = createFontList(self.ttffiles)
  File "/Users/Apollo/venv2714/lib/python2.7/site-packages/matplotlib/font_manager.py", line 587, in createFontList
    font = ft2font.FT2Font(fpath)
TypeError: First argument must be a path or file object reading bytes

I encounterd no problem with version 2.2.0

`Mac-Pro-di-Apollo:~ Apollo$ source venv2714/bin/activate`
`(venv2714)Mac-Pro-di-Apollo:~ Apollo$ python`
`Python 2.7.9 (v2.7.9:648dcafa7e5f, Dec 10 2014, 10:10:46) `
`[GCC 4.2.1 (Apple Inc. build 5666) (dot 3)] on darwin`

`import matplotlib`
`matplotlib.use('TkAgg')`
`import matplotlib.pyplot`
`print matplotlib.__version__`
`2.2.0`

**Comment by lucilla7** [comment]
No problem also with version 2.2.0 and Python 2.7.1.

It seems like I have problems only with version 2.2.2

**Comment by tacaswell** [comment]
hmm, that points to a different change than we originally blamed this on.

@lucilla7 can you try building the 2.2.x branch locally to make sure it works for you?

**Comment by lucilla7** [comment]
Yes it works:

Processing /Users/Apollo/Downloads/matplotlib-2.2.x
Requirement already satisfied: numpy>=1.7.1 in /Users/Apollo/venv/lib/python2.7/site-packages (from matplotlib===2.2.x)
Requirement already satisfied: cycler>=0.10 in /Users/Apollo/venv/lib/python2.7/site-packages (from matplotlib===2.2.x)
Requirement already satisfied: pyparsing!=2.0.4,!=2.1.2,!=2.1.6,>=2.0.1 in /Users/Apollo/venv/lib/python2.7/site-packages (from matplotlib===2.2.x)
Requirement already satisfied: python-dateutil>=2.1 in /Users/Apollo/venv/lib/python2.7/site-packages (from matplotlib===2.2.x)
Requirement already satisfied: pytz in /Users/Apollo/venv/lib/python2.7/site-packages (from matplotlib===2.2.x)
Requirement already satisfied: six>=1.10 in /Users/Apollo/venv/lib/python2.7/site-packages (from matplotlib===2.2.x)
Requirement already satisfied: kiwisolver>=1.0.1 in /Users/Apollo/venv/lib/python2.7/site-packages (from matplotlib===2.2.x)
Requirement already satisfied: backports.functools_lru_cache in /Users/Apollo/venv/lib/python2.7/site-packages (from matplotlib===2.2.x)
Requirement already satisfied: subprocess32 in /Users/Apollo/venv/lib/python2.7/site-packages (from matplotlib===2.2.x)
Requirement already satisfied: setuptools in /Users/Apollo/venv/lib/python2.7/site-packages (from kiwisolver>=1.0.1->matplotlib===2.2.x)
Building wheels for collected packages: matplotlib
  Running setup.py bdist_wheel for matplotlib ... done
  Stored in directory: /Users/Apollo/Library/Caches/pip/wheels/f2/68/7e/041429c5aa316caa80e28442dd032b7ae05f5bdb27e3922d72
Successfully built matplotlib
Installing collected packages: matplotlib
  Found existing installation: matplotlib 2.2.0
    Uninstalling matplotlib-2.2.0:
      Successfully uninstalled matplotlib-2.2.0
Successfully installed matplotlib-2.2.x

**Comment by tacaswell** [comment]
Great!  I'm going to close this as resolved and tag 2.2.3 tonight (independent if anything else pops up!)
